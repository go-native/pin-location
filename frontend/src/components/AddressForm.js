import React, { Component } from "react"
import Autosuggest from "react-autosuggest"

const getSuggestionValue = suggestion => suggestion.address
const renderSuggestion = suggestion => <div>{suggestion.address}</div>

const debounce = (func, delay) => {
  let inDebounce
  return function() {
    const context = this
    const args = arguments
    clearTimeout(inDebounce)
    inDebounce = setTimeout(() => func.apply(context, args), delay)
  }
}

class AddressForm extends Component {
  state = this.props.addressToEdit ? {...this.props.addressToEdit}
    : { title: "", address: "" }
  submitForm = event => {
    event.preventDefault()
    this.isFormValid() && this.props.onSaveAddress(this.state)
  }
  debounceSearchAddress = debounce(value => {
    this.props.onSearchAddress(value);
  }, 500)
  onSuggestionsFetchRequested = e => {
    this.debounceSearchAddress(e.value);
  }
  onSuggestionsClearRequested = () => {
    this.props.onFlushAvailableAddresses();
  }
  onAddressChange = (event, { newValue }) => {
    this.setState({
      address: newValue,
      lat: null,
      lng: null
    })
  }
  handleTitleChange = event => {
    this.setState({title: event.target.value})
  }
  onSuggestionSelected = (event, { suggestion }) => {
    this.setState({...suggestion})
  }
  isFormValid() {
    return this.state.title.trim().length !== 0 && this.state.address.trim().length !== 0
      && this.state.lat && this.state.lng
  }
  render() {
    const inputProps = {
      placeholder: "Query geocoding API",
      value: this.state.address,
      onChange: this.onAddressChange
    }
    return (
      <form className="form" onSubmit={this.submitForm}>
        <div className="label">Title:</div>
        <input
          type="text"
          className="text_field"
          value={this.state.title}
          onChange={this.handleTitleChange}
        />
        <div className="label">Address:</div>
        <Autosuggest
          suggestions={this.props.availableAddresses}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          getSuggestionValue={getSuggestionValue}
          renderSuggestion={renderSuggestion}
          onSuggestionSelected={this.onSuggestionSelected}
          inputProps={inputProps}
        />
        {this.state.lat && this.state.lng && (
          <div className="label">
            Latitude: {this.state.lat} / Longitude: {this.state.lng}
          </div>
        )}
        <br />
        <div>
          <input className="action-btn save-btn" type="submit" value="Submit"  disabled={!this.isFormValid()}/>
          <button
            className="action-btn cancel-btn"
            onClick={this.props.onCancel}
          >
            Cancel
          </button>
        </div>
      </form>
    );
  }
}

export default AddressForm
