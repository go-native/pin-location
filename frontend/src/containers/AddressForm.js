import React, { Component } from "react"
import { connect } from "react-redux"
import AddressForm from "../components/AddressForm"
import {
  closeForm,
  searchAddress,
  saveAddress,
  flushAvailableAddresses
} from "../actions"

class AddressFormContainer extends Component {
  render() {
    return (
      <div className={`side-container ${this.props.isFormOpened && "side-container--open"}`}>
        {this.props.isFormOpened && <AddressForm {...this.props} />}
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    isFormOpened: state.isFormOpened,
    availableAddresses: state.availableAddresses,
    addressToEdit: state.addressToEdit
  }
}
const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    onCancel: () => dispatch(closeForm()),
    onSearchAddress: name => dispatch(searchAddress(name)),
    onFlushAvailableAddresses: () => dispatch(flushAvailableAddresses()),
    onSaveAddress: address => dispatch(saveAddress(address))
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddressFormContainer)
