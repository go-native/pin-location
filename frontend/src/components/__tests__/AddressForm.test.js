import React from "react"
import { shallow } from "enzyme"
import AddressForm from "../AddressForm";

const address = {
  id: "1544109127653",
  title: "Leipziger str.",
  address: "Leipziger strasse",
  lat: 52.511123,
  lng: 13.399605
}

describe("AddressForm", () => {
  it("should render properly for creating", () => {
    const wrapper = shallow(<AddressForm isFormOpened={true} availableAddresses={[]}/>)
    expect(wrapper).toMatchSnapshot()
  })
  it("should render properly for editing", () => {
    const wrapper = shallow(<AddressForm isFormOpened={true} availableAddresses={[]} addressToEdit={address}/>)
    expect(wrapper).toMatchSnapshot()
  })
  it("should be not valid when any input is missing", () => {
    const wrapper = shallow(<AddressForm isFormOpened={true} availableAddresses={[]}/>)
    const submitBtnElem = wrapper.find('.save-btn')
    expect(submitBtnElem.props().disabled).toBe(true)
  })
  it("should set the state properly", () => {
    const wrapper = shallow(<AddressForm isFormOpened={true} availableAddresses={[]}/>)

    const titleTextElem = wrapper.find('.text_field')
    titleTextElem.simulate('change', {target: {value: "Test"}})
    expect(wrapper.state().title).toBe("Test")

    wrapper.instance().onAddressChange('event', {newValue: address.address})
    expect(wrapper.state().address).toBe(address.address)

    wrapper.instance().onSuggestionSelected('event', {suggestion: {address: address.address, lat: address.lat, lng: address.lng}})
    expect(wrapper.state().address).toBe(address.address)
    expect(wrapper.state().lat).toBe(address.lat)
    expect(wrapper.state().lng).toBe(address.lng)
  })
  it("should be valid when all inputs are set", () => {
    const wrapper = shallow(<AddressForm isFormOpened={true} availableAddresses={[]}/>)
    wrapper.setState(address)
    const submitBtnElem = wrapper.find('.save-btn')
    expect(submitBtnElem.props().disabled).toBe(false)
  })
  it("should call proper function when submitted and pass arguments", () => {
    const onSaveClick = jest.fn()
    const wrapper = shallow(<AddressForm isFormOpened={true} onSaveAddress={onSaveClick}  availableAddresses={[]}/>)
    const formElem = wrapper.find('.form')
    wrapper.setState(address)
    formElem.simulate('submit', {preventDefault: jest.fn()})
    expect(onSaveClick).toHaveBeenCalledTimes(1)
    expect(onSaveClick).toHaveBeenCalledWith(address)
  })
  it("it should call proper function when canceled", () => {
    const onCancelClick = jest.fn()
    const wrapper = shallow(<AddressForm isFormOpened={true} onCancel={onCancelClick}  availableAddresses={[]}/>)
    const cancelBtnElem = wrapper.find('.cancel-btn')
    cancelBtnElem.simulate('click')
    expect(onCancelClick).toHaveBeenCalledTimes(1)
  })
})