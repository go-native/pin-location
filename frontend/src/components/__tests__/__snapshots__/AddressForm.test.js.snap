// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm should render properly for creating 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="label"
  >
    Title:
  </div>
  <input
    className="text_field"
    onChange={[Function]}
    type="text"
    value=""
  />
  <div
    className="label"
  >
    Address:
  </div>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Query geocoding API",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  />
  <br />
  <div>
    <input
      className="action-btn save-btn"
      disabled={true}
      type="submit"
      value="Submit"
    />
    <button
      className="action-btn cancel-btn"
    >
      Cancel
    </button>
  </div>
</form>
`;

exports[`AddressForm should render properly for editing 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="label"
  >
    Title:
  </div>
  <input
    className="text_field"
    onChange={[Function]}
    type="text"
    value="Leipziger str."
  />
  <div
    className="label"
  >
    Address:
  </div>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Query geocoding API",
        "value": "Leipziger strasse",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  />
  <div
    className="label"
  >
    Latitude: 
    52.511123
     / Longitude: 
    13.399605
  </div>
  <br />
  <div>
    <input
      className="action-btn save-btn"
      disabled={false}
      type="submit"
      value="Submit"
    />
    <button
      className="action-btn cancel-btn"
    >
      Cancel
    </button>
  </div>
</form>
`;
